### ISSUE TITLE
HFS+ on Yosemite (10.10.2) can't icat files "Attribute 4352 not found", possibly new compression algorithm

### ISSUE THREAD

**Issue by destijl** [sleuthkit/sleuthkit#401]
TL;DR I think resource fork compression changed in HFS+ on Yosemite.  I tested with 4.1.3 release and the develop HEAD.  Same results.

On 10.9.5 things work great:

> $ sw_vers 
> ProductName:  Mac OS X
> ProductVersion:   10.9.5
> BuildVersion: 13F34`

Note the "Data is zlib compressed in the resource fork" line from this istat of /bin/ls:

> $ sudo istat /dev/rdisk1 11571
> File Path: /bin/ls
> Catalog Record: 11571
> Allocated
> Type: File
> Mode: rrwxr-xr-x
> Size: 34736
> uid / gid: 0 / 0
> Link count:   1
> 
> File Name: ls
> Admin flags: 0
> Owner flags: 32 - compressed
> Has extended attributes
> File type:    0000  
> File creator: 0000  
> Text encoding:        0 = MacRoman
> Resource fork size:   15290
> 
> Times:
> Created:      2013-12-19 15:19:45 (PST)
> Content Modified:     2014-05-29 06:06:32 (PDT)
> Attributes Modified:  2014-05-29 06:06:32 (PDT)
> Accessed:     2015-03-17 16:46:49 (PDT)
> Backed Up:    0000-00-00 00:00:00 (UTC)
> 
> Resource Fork Blocks:
> 414451-414454
> 
> Attributes:
> Type: CMPF (4355-2)   Name: com.apple.decmpfs   Resident   size: 16
> Type: RSRC (4353-1)   Name: RSRC   Non-Resident   size: 15290  init_size: 15290
> Type: DATA (4352-0)   Name: DATA   Non-Resident, Compressed   size: 15290  init_size: 15290
> 
> Compressed File:
>     Uncompressed size: 34736
>     Data is zlib compressed in the resource fork
> 
> Resources:
>   Type: cmpf  ID: 1           Offset: 260     Size: 14980     Name: <none>

On 10.10.2 I suspect there is a different compression scheme:

> $ sw_vers
> ProductName:  Mac OS X
> ProductVersion:       10.10.2
> BuildVersion: 14C109

Note compression is now unknown, doing the istat on /bin/ls again:

> $ sudo istat /dev/rdisk1 1316086
> File Path: /bin/ls
> Catalog Record: 1316086
> Allocated
> Type:   File
> Mode:   rrwxr-xr-x
> Size:   34640
> uid / gid: 0 / 0
> Link count:     1
> 
> File Name: ls
> Admin flags: 0
> Owner flags: 32 - compressed
> Has extended attributes
> File type:      0000  
> File creator:   0000  
> Text encoding:  0 = MacRoman
> Resource fork size:     16389
> 
> Times:
> Created:        2014-09-09 15:49:59 (PDT)
> Content Modified:       2014-09-09 15:49:59 (PDT)
> Attributes Modified:    2015-01-23 18:02:35 (PST)
> Accessed:       2015-03-17 16:57:29 (PDT)
> Backed Up:      0000-00-00 00:00:00 (UTC)
> 
> Resource Fork Blocks:
> 5397171-5397175
> 
> Attributes:
> Type: CMPF (4355-2)   Name: com.apple.decmpfs   Resident   size: 16
> Type: RSRC (4353-1)   Name: RSRC   Non-Resident   size: 16389  init_size: 16389
> 
> Compressed File:
>     Uncompressed size: 34640
>     Compression type is UNKNOWN

I can actually get the compressed data out fine:

> $ sudo icat /dev/rdisk1 1316086-4353-1 > temp.blah

I messed around with that data a bit but couldn't come up with what scheme is used, there isn't a header that seemed to match with the common formats I looked for. Here's the first few bytes, you can see the Mach-O header starting with the FEED FACF magic string.

> 08 00 00 00  05 40 00 00  E0 01 CF FA  ED FE

**Comment by destijl** [comment]
Should have mentioned, the error you get is "Attribute 4352 not found":

$ sudo icat /dev/rdisk1 1316086                   
Attribute not found in file (tsk_fs_attrlist_get: Attribute 4352 not found)

**Comment by destijl** [comment]
FYI I tested today's head with @meder's HFS+ fixes, but they don't address this problem.

**Comment by destijl** [comment]
Looks like it's LZVN:
http://reverseengineering.stackexchange.com/questions/8209/unknown-mac-osx-10-10-hfs-compression
https://pikeralpha.wordpress.com/2014/10/09/lzvn-packed-apple-logos/

code here:
https://github.com/Piker-Alpha/LZVN

**Comment by dimkr** [comment]
It seems cases 7-10 of compression_type from http://www.opensource.apple.com/source/copyfile/copyfile-118.1.2/copyfile.c need to be added to hfs.c. I added them and now I get this in icat, when I specify the inode:
```
Invalid file offset (tsk_fs_attr_read - 738197508) ( hfs_attr_walk_special: trying to read the offset table size, return value of 4294967295 should have been 4)
```

Some function returns -1 and I don't know which one or why.

**Comment by joachimmetz** [comment]
@dimkr what "clues" are you after?

> Some function returns -1 and I don't know which one or why.

You'll need to be more specific, which functions?

> Although ffind and friends seem to work fine now

Does ffind need to read the file content data? As the original report indicates, the issue is with icat.

**Comment by dimkr** [comment]
> You'll need to be more specific, which functions?

tsk_fs_attr_read() returns -1 and hfs_attr_walk_special() fails because of this.

**Comment by joachimmetz** [comment]
The error I encounter is ```tsk_fs_attrlist_get: Attribute 4352 not found``` generated here: https://github.com/sleuthkit/sleuthkit/blob/d2027d6db94617b080461baa8e093470e000a020/tsk/fs/fs_attrlist.c#L243

> Although ffind and friends seem to work fine now

Which version of TSK ? 4.2.0 stable or the development branch?

**Comment by hexadite-dima** [comment]
4.2.0.

-- 
The information in this e-mail transmission contains proprietary and 
business sensitive information.  Unauthorized interception of this e-mail 
may constitute a violation of law. If you are not the intended recipient, 
you are hereby notified that any review, dissemination, distribution or 
duplication of this communication is strictly prohibited. You are also 
asked to contact the sender by reply email and immediately destroy all 
copies of the original message.

**Comment by andreas-eriksson** [comment]
I encountered the same issue recently. Has there been any progress on this?

**Comment by MalfurionStormrage** [comment]
I'm stuck on the same problem. I've acquired an HFS+ and I get this error when run "icat" against certain files:
icat -i raw -f hfs H:\IFOM02\ifom02.dd 938309
Attribute not found in file (tsk_fs_attrlist_get: Attribute 4352 not found)

If I mount the same image through Mac OSx (yosemite/captain/sierra) I can see those files correctly.

**Comment by uckelman-sf** [comment]
We have LZVN support completed now, fwiw.

**Comment by uckelman-sf** [comment]
This is fixed in PR #830.

**Comment by uckelman-sf** [comment]
@simsong This should be closed.
